[[ Bug 21982 ]] Ensure 'the printersettings' return the correct value

livecodepanos opened the pull request:
This patch fixes an error in pointer arithmetic, that resulted in the chosen `printersettings` not being found, thus default values were used instead.

Closes https://quality.livecode.com/show_bug.cgi?id=21982

Tested on Windows 7.

livecodeian commented on engine/src/w32printer.cpp:
```diff
@@ -1986,7 +1986,7 @@ bool MCWindowsPrinter::FetchDialogData(HGLOBAL& r_devmode_handle, HGLOBAL& r_dev
 
 			// SN-2014-08-07: [[ Bug 13084 ]] The pointer arithmetic wasn't right
-			MCStringGetChars(*t_string, MCRangeMake(0, MCStringGetLength(*t_string)), (unichar_t*)(t_devnames + 1));
+			MCStringGetChars(*t_string, MCRangeMake(0, MCStringGetLength(*t_string)), (unichar_t*)(t_devnames + sizeof(unichar_t) * 1));
```
this doesn't look right - the old code was adding 1 to t_devnames which is a pointer to DEVNAME resulting (in bytes) in t_devname + 1 * sizeof(DEVNAME). From the docs https://docs.microsoft.com/en-us/windows/win32/api/commdlg/ns-commdlg-devnames this looks like the correct thing to do - paste the string after the struct.
I think the error is in the values added to wDriverOffset and wOutputOffset which don't take the size of unichar_t into account
```
t_devnames -> wDriverOffset = t_devnames_size - 1 * sizeof(unichar_t);
...
t_devnames -> wOutputOffset = GetDeviceOutputType() == PRINTER_OUTPUT_FILE ? t_devnames_size - 6 * sizeof(unichar_t) : t_devnames_size - 1 * sizeof(unichar_t);
```

livecodepanos replied:
@livecodeian Hmm I think I had tried the suggested change and it did not work - let me try this again.

livecodepanos replied:
Yeap, the suggested change does not seem to fix the problem.

Recipe:
```
on mouseUp
   answer printer // choose a printer other than the default
   answer page setup
   answer the printerName // --> the default printer is returned instead of the one chosen before
end mouseUp
```

livecodeian replied:
There are similar errors in StoreDialogData handling of DEVNAMES which may be why it's still broken

livecodepanos replied:
@livecodeian where exactly in StoreDialogData handling of DEVNAMES? I cannot spot anything wrong

livecodeian replied:
Actually it's fine as is - I misread the API docs. The offsets in the struct are given in chars rather than bytes so what needs to change is the calculations involving `sizeof(DEVNAMES)` as that returns a number of bytes so:
Line 1977-1978 becomes
```
t_devnames_size = sizeof(DEVNAMES) / sizeof(unichar_t) + MCStringGetLength(*t_string);
t_devnames_handle = GlobalAlloc(GMEM_MOVEABLE, t_devnames_size * sizeof(unichar_t));
```
Line 1985 becomes
`t_devnames -> wDeviceOffset = sizeof(DEVNAMES) / sizeof(unichar_t);`

livecodepanos replied:
Thanks @livecodeian, that makes sense. This change returns correct results with most of the printers. However, there are still printers where if I choose them in `answer printer`, and then call `answer page setup`, they are not saved properly, and calling `answer the printername` returns the default printer instead of the chosen one in these cases. Choosing the same printers works fine in LC 6.x, so it seems that still something somewhere is broken :(

livecodeian replied:
Ah! I notice the string construction on line 1971 uses a c-string literal with escaped nulls (`'\0'`) as a format string. That won't work as the formatting will stop at the first null encountered. Instead try constructing a mutable string with MCStringAppend:
```
MCStringCreateMutable(&t_string);
MCStringAppend(*t_string, m_name);
MCStringAppendChar(*t_string, '\0');
MCStringAppend(*t_string, MCSTR("FILE:"));
MCStringAppendChar(*t_string, '\0');

livecodepanos replied:
This sounds promising @livecodeian :)

I'll try this tomorrow - thanks!

livecodepanos replied:
Works like a charm, good catch @livecodeian :)

livecodepanos committed:
[[ Bug 21982 ]] Ensure 'the printersettings' return the correct value

This patch fixes an error in pointer arithmetic, that resulted in the
chosen printersettings not being found, thus default values were used instead.

livecodeian commented:
@livecode-vulcan review ok 2252f84

livecode-vulcan commented:
💙 review by @livecodeian ok 2252f84fcb1fa76ef622ee75e8732085dbec0ee7

livecode-vulcan commented:
😎 test success 2252f84fcb1fa76ef622ee75e8732085dbec0ee7

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/973)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/381)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/380)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/379)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/380)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2498)
* try-community-universal-ios-iphoneos14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.4/builds/27)
* try-community-universal-ios-iphoneos14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.1/builds/76)
* try-community-universal-ios-iphonesimulator14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.4/builds/27)
* try-community-universal-ios-iphonesimulator14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.1/builds/74)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2525)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2514)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2512)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2549)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2485)
